src/JsonSchema/Constraints/TypeConstraint.php

```diff
@@ -41,48 +41,99 @@ class TypeConstraint extends Constraint
     public function check($value = null, $schema = null, $path = null, $i = null)
     {
         $type = isset($schema->type) ? $schema->type : null;
-        $isValid = true;
+        $isValid = false;
+        $wording = [];
 
         if (is_array($type)) {
-            // @TODO refactor
-            $validatedOneType = false;
-            $errors = array();
-            foreach ($type as $tp) {
-                $validator = new static($this->checkMode);
-                $subSchema = new \stdClass();
-                $subSchema->type = $tp;
-                $validator->check($value, $subSchema, $path, null);
-                $error = $validator->getErrors();
-
-                if (!count($error)) {
-                    $validatedOneType = true;
-                    break;
-                }
-
-                $errors = $error;
-            }
-
-            if (!$validatedOneType) {
-                $this->addErrors($errors);
-
-                return;
-            }
+            $this->validateTypesArray($value, $type, $wording, $isValid, $path);
         } elseif (is_object($type)) {
             $this->checkUndefined($value, $type, $path);
+            return;
         } else {
             $isValid = $this->validateType($value, $type);
         }
 
         if ($isValid === false) {
-            if (!isset(self::$wording[$type])) {
-                throw new StandardUnexpectedValueException(
-                    sprintf(
-                        "No wording for %s available, expected wordings are: [%s]",
-                        var_export($type, true),
-                        implode(', ', array_filter(self::$wording)))
-                );
+            if (!is_array($type)) {
+                $this->validateTypeNameWording($type);
+                $wording[] = self::$wording[$type];
+            }
+            $this->addError($path, ucwords(gettype($value)) . " value found, but " .
+                $this->implodeWith($wording, 'or') . " is required", 'type');
+        }
+    }
+
+    /**
+     * Validates the given $value against the array of types in $type. Sets the value
+     * of $isValid to true, if at least one $type mateches the type of $value or the value
+     * passed as $isValid is already true.
+     *
+     * @param mixed $value Value to validate
+     * @param array $type TypeConstraints to check agains
+     * @param array $wording An array of wordings of the valid types of the array $type
+     * @param boolean $isValid The current validation value
+     */
+    protected function validateTypesArray($value, array $type, &$validTypesWording, &$isValid,
+                                          $path) {
+        foreach ($type as $tp) {
+            // $tp can be an object, if it's a schema instead of a simple type, validate it
+            // with a new type constraint
+            if (is_object($tp)) {
+                if (!$isValid) {
+                    $validator = new static($this->checkMode);
```

Comment:
I thought we could get rid of this (at least for me) ugly looking workaround, but it seems it's the best solution we currently have :/